### ISSUE TITLE
MTP speculative decoding diverges from autoregressive output at temperature 0 on quantized Gemma 4 12B

### ISSUE THREAD

**Issue by bekla050** [ml-explore/mlx-swift-lm#542]
## Summary

On quantized `gemma4_unified` (Gemma 4 12B) targets, MTP speculative decoding produces different
text than autoregressive generation at `temperature: 0`, once generation passes roughly 200 tokens.
This breaks the bit-exact equivalence that `testMTP31BMatchesBaselineByteIdentical` asserts, and
that speculative decoding depends on to be a pure speedup rather than a change in behavior.

Speculation itself is healthy in every run below: no passthrough, acceptance 29–84 %, `stopReason`
is `stop`, and output length matches baseline exactly. The output does not decay into repetition —
both texts are coherent, they are simply different.

The existing byte-identity tests run against the 31B bf16/8-bit pair at `maxTokens: 32`. Quantized
12B targets are not covered, and 32 tokens is short enough that the divergence has usually not
appeared yet.

## Environment

- mlx-swift-lm at `d7dc03d8447ee6b42b54a1c5295b4e56ee9274f3` (main, 2026-08-14)
- mlx-swift 0.31.6
- macOS 26, Apple Silicon, 128 GB
- Target loaded through `VLMModelFactory` (`gemma4_unified`); drafter
  `mlx-community/gemma-4-12B-it-qat-assistant-4bit`, instantiated directly as
  `Gemma4AssistantDraftModel(config)` + `loadWeights(modelDirectory:model:quantization:perLayerQuantization:)`
- `blockSize: 4`, `temperature: 0`, no repetition or presence penalties
- No KV cache quantization (`kvBits` and `maxKVSize` both unset)

## Reproduction

Both arms use the same target, the same `LMInput` (built once via
`context.processor.prepare(input:)`), and the same free `generate(...)` entry point. The only
difference is the `mtpDrafter:` argument.

```swift
// Baseline
let baseline = try generate(
    input: lmInput, parameters: GenerateParameters(maxTokens: n, temperature: 0),
    context: context)

// MTP
let speculative = try generate(
    input: lmInput, parameters: GenerateParameters(maxTokens: n, temperature: 0),
    context: context, mtpDrafter: drafter, blockSize: 4)
```

The prompt is a summarization instruction over a long synthetic German meeting transcript,
generated deterministically by this script (~126 KB of text, 28,927 prompt tokens). A long,
content-rich prompt matters — see "Prompt length matters" below.

```python
import random
random.seed(20260815)
names = ["Alex", "Robin", "Kim", "Sam", "Noa", "Luca"]
topics = ["Rechnungslauf", "Wochenbericht", "Urlaubsplanung", "Serverumzug", "Schulungstermine",
          "Lieferantenwechsel", "Budgetfreigabe", "Ticketsystem", "Onboarding", "Zeiterfassung"]
verbs = ["schlägt vor", "fragt nach", "widerspricht bei", "notiert zu", "übernimmt",
         "verschiebt", "bestätigt", "hinterfragt", "dokumentiert", "priorisiert"]
tails = ["bis Ende der Woche", "nach Rücksprache mit der Fachabteilung", "im nächsten Sprint",
         "sobald die Freigabe vorliegt", "vor dem Quartalsabschluss", "in der Folgesitzung"]
lines = []
for _ in range(900):
    s, t, v, x = (random.choice(names), random.choice(topics),
                  random.choice(verbs), random.choice(tails))
    lines.append(f"{s}: Zum Thema {t} — ich {v.split()[0]} das {x}. "
                 f"Wir sollten den Punkt {t} nicht länger offen lassen.")
transcript = "\n".join(lines)
```

The transcript is then embedded in a plain summarization request (system prompt: a protocol-writing
instruction; user prompt: the transcript wrapped in a task block). Any equivalent long German
summarization prompt should work.

## Result — synthetic prompt, `mlx-community/gemma-4-12B-it-qat-4bit`

Each row is a separate run at a different `maxTokens`. "identical" compares the MTP output to the
baseline output byte for byte.

| `maxTokens` | tokens baseline/MTP | identical | first divergence | acceptance |
|---|---|---|---|---|
| 32 | 32 / 32 | yes | — | 69.0 % |
| 64 | 64 / 64 | yes | — | 69.5 % |
| 128 | 128 / 128 | yes | — | 74.1 % |
| 256 | 256 / 256 | **no** | char 717 | 74.2 % |
| 512 | 512 / 512 | **no** | char 717 | 80.2 % |
| 1024 | 1024 / 1024 | **no** | char 717 | 83.8 % |

`passthroughReason` is nil and `proposedDraftTokens > 0` in every row. Repeating any row reproduces
the same divergence offset byte for byte, and baseline runs are self-identical across repetitions,
so this is not baseline nondeterminism.

## The onset depends on the quantization scheme, but every quantization diverges

Same setup, same drafter, three independently produced 4-bit quantizations of the same model, on a
different (longer-output) German summarization prompt:

| Target | 32 | 64 | 128 | 256 | 512 | first divergence | quantization |
|---|---|---|---|---|---|---|---|
| `mlx-community/gemma-4-12B-it-qat-OptiQ-4bit` | yes | yes | yes | **no** | **no** | char 782 | mixed 4/8-bit, 329 per-layer entries |
| `lmstudio-community/gemma-4-12B-it-MLX-4bit` | yes | yes | yes | **no** | **no** | char 619 | uniform 4-bit, no per-layer entries |
| `mlx-community/gemma-4-12B-it-qat-4bit` | **no** | **no** | **no** | **no** | **no** | char 27 | mixed, 144 per-layer entries |

On that prompt the third target diverges within the first 32 generated tokens — the same length the
existing byte-identity regression test uses.

## Prompt length matters

With a short prompt (49 tokens, 16 generated tokens) MTP output is byte-identical to baseline on all
three targets. The divergence only appears with long prompts and longer generations, which is why
the current 32-token test on a short prompt does not catch it.

## What we ruled out

- **Passthrough.** `passthroughReason` is nil and `proposedDraftTokens > 0` in every run reported
  here. We verified that our detection of passthrough actually works by deliberately loading the
  target through `LLMModelFactory` instead: that yields `"main model did not emit drafter state"`,
  zero proposals, and — by construction — output identical to baseline.
- **Baseline nondeterminism.** Baseline repetitions are byte-identical to each other at this SHA.
- **Model weights.** Three independently produced quantizations diverge, including a uniform 4-bit
  one with no per-layer overrides.
- **KV cache quantization.** Not enabled in any run.
- **Prompt assembly.** Both arms consume the identical `LMInput`; `promptTokenCount` matches.

## What we did not test

- A bf16 or 8-bit 12B target — we only have quantized 12B weights locally. If the divergence
  disappears there, quantization noise around near-tie logits during the batched verify pass is the
  likely mechanism, and the right fix might be test coverage for quantized targets plus a
  documented caveat rather than a code change.
- Whether output quality differs systematically. We only established that the outputs differ.

## Suggestion

Extend byte-identity coverage to at least one quantized target with a long prompt and a generation
length past ~256 tokens. Based on the table above, `mlx-community/gemma-4-12B-it-qat-4bit`
reproduces this without needing the 35 GB 31B pair.

**Comment by aleroot** [comment]
This is not a problem of this library per se, this is a current limitation of MLX. 
Single-token autoregressive decoding and multi-token speculative verification can select different Metal matmul and attention kernels. Their different floating-point reduction orders produce small BF16 differences that may eventually flip an argmax, particularly with long contexts and quantized models.

This also affects Qwen3.5 MTP and other speculative-decoding implementations using MLX, thanks @bekla050 you have found something really interesting... 

The [fix](https://github.com/ml-explore/mlx/pull/4279) addresses the numerical discrepancy at its source instead of adding model-specific tolerances or serially re-verifying speculative tokens.

Also opened https://github.com/ml-explore/mlx-swift/pull/452 and https://github.com/ml-explore/mlx-c/pull/124

**Comment by aleroot** [comment]
It looks like Apple does not want to support this, they prefer user with those requirements to move elsewhere, @bekla050 I guess you will need to close the ticket, as they won't accept.
